src/install-pnpm/run.ts

```diff
@@ -37,12 +37,78 @@ export async function runSelfInstaller(inputs: Inputs): Promise<number> {
   return exitCode
 }
 
+// Nearly identical to the function `actions/setup-node` uses.
+// See https://github.com/actions/setup-node/blob/39370e3970a6d050c480ffad4ff0ed4d3fdee5af/src/util.ts#L8
+function getPnpmVersionFromFile(versionFilePath: string) {
+  if (!existsSync(versionFilePath)) {
+    throw new Error(
+      `The specified pnpm version file at: ${versionFilePath} does not exist`
+    )
+  }
+
+  const contents = readFileSync(versionFilePath, 'utf8')
+
+  // Try parsing the file as a `package.json` file.
+  try {
+    const manifest = JSON.parse(contents)
+
+    // Presume package.json file.
+    if (typeof manifest === 'object' && !!manifest) {
+      // Support Volta.
+      // See https://docs.volta.sh/guide/understanding#managing-your-project
+      if (manifest.volta?.pnpm) {
+        return manifest.volta.pnpm as string
+      }
+
+      if (manifest.engines?.pnpm) {
+        return manifest.engines.pnpm as string
+      }
+
+      // Support Volta workspaces.
+      // See https://docs.volta.sh/advanced/workspaces
+      if (manifest.volta?.extends) {
+        const extendedFilePath = path.resolve(
+          path.dirname(versionFilePath),
+          manifest.volta.extends
+        )
+        console.info('Resolving pnpm version from ' + extendedFilePath)
+        return getPnpmVersionFromFile(extendedFilePath)
+      }
+
+      // If contents are an object, we parsed JSON
+      // this can happen if pnpm-version-file is a package.json
+      // yet contains no volta.pnpm or engines.pnpm
+      //
+      // If pnpm-version file is _not_ JSON, control flow
+      // will not have reached these lines.
+      //
+      // And because we've reached here, we know the contents
+      // *are* JSON, so no further string parsing makes sense.
+      throw new Error(`${versionFilePath} was supplied for 'version_file_path', but does not contain a pnpm version in a recognized property.`)
+    }
+  } catch {
+    console.info('pnpm version file is not JSON file, trying to parse as text')
+  }
+
+  // If not JSON, try and grab a version from a few possible text formats.
+  const matched = contents.match(/^(?:pnpm\s+)?v?(?<version>[^\s]+)$/m)
+  const found = matched?.groups?.version ?? contents.trim()
```

Comment:
Regex is overkill here.

The format is simple enough. Just use `.split('\n')`, `.find()`, `.startsWith('pnpm')`, and `.slice('pnpm'.length).trim()`.